### PR TITLE
FindNext/FindPrevious: fix selection overflow crash, off-by-one, and overlapping regex matches

### PR DESCRIPTION
**Pull request by mjuhasz**
## Summary

Fixes a cluster of related bugs in the Find/FindNext/FindPrevious flow that caused crashes, missed matches, and incorrect selections.

## Motivation

Four separate issues were found when using Find with AvaloniaEdit:

1. **Crash on selection overflow** — AvaloniaEdit's `SelectionStart` setter internally calls `Select(value, currentSelectionLength)`, which throws an `ArgumentOutOfRangeException` when the new start plus the existing selection length exceeds the document length (e.g. navigating from a long match to a short one).

2. **FindNext keyboard shortcut skips first character** — The shortcut was passing `currentCharIndex + 1` to `FindNext`, causing a one-character over-advance that made matches starting at the current caret position unreachable.

3. **Stale text guard too narrow** — The text-box sync check before setting a selection was `if (EditTextBox.Text == string.Empty)`, which missed the case where a stale subtitle (not empty, just wrong) was still displayed when the dispatcher callback ran.

4. **FindPrevious misses overlapping regex matches** — `FindWithRegexReverse` used `Regex.Matches()`, which does not yield overlapping results. A regex like `(^|\n).{38,}($|\n)` on a two-line subtitle produces two matches sharing the `\n` boundary; only the first was found, so traversing backwards would jump over the second-line match entirely.

### `TextEditorWrapper.cs`
- Added a new `Select(int start, int length)` helper that clears the selection length before repositioning, avoiding the AvaloniaEdit overflow throw.
- Caret is placed at the end of the selection (consistent with standard find-highlight UX).

### `MainViewModel.cs`
- Replaced all four occurrences of the three-line selection pattern (`CaretIndex = x; SelectionStart = x; SelectionEnd = x + len`) with a single `EditTextBox.Select(index, length)` call.
- Fixed the `+ 1` off-by-one in the FindNext keyboard shortcut.
- Changed the text-box sync guard from `== string.Empty` to `!= subtitle.Text` so stale (not just empty) content is overwritten before a selection is applied.

### `FindService.cs`
- Rewrote `FindWithRegexReverse` to iterate through matches with a step of 1 after each hit (`pos = m.Index + 1`), ensuring overlapping matches sharing a boundary character are all found. The last match within the search prefix is returned, preserving the reverse-traversal semantic.

### `FindServiceTests.cs`
- Added `RegexOverlappingNewlineBoundary_FindNext_FindsSecondLineInFull` — verifies that FindNext advances past the first overlapping match and finds the second line's match in full.
- Added `RegexOverlappingNewlineBoundary_FindPrevious_FindsSecondLineBeforeFirst` — verifies that FindPrevious traversing from the end finds the second-line match before the first-line match.

## Testing

All existing tests pass. New regression tests cover the overlapping-boundary case in both directions.

### CODE REVIEW

**Comment by mjuhasz** [comment]
Let me look into these, the feedback is valid. I think, rather than patching the call site to treat the symptom, I can fix the root-case with a one-liner only.

**Comment by niksedk** [comment]
> Let me look into these, the feedback is valid. I think, rather than patching the call site to treat the symptom, I can fix the root-case with a one-liner only.



> Let me look into these, the feedback is valid. I think, rather than patching the call site to treat the symptom, I can fix the root-case with a one-liner only.

Yes, Find/replace is more complicated that it sounds ;)

It looks good to me....

**Comment by mjuhasz** [comment]
> It looks good to me....

To me as well, with or without the suggested change, it works the same way, nevertheless I made the fix for good hygiene.

TextBoxWrapper.Select() was indeed inconsistent. It wasn't updating CaretIndex while TextEditorWrapper.Select() explicitly sets CaretOffset = start + length.

I've fixed TextBoxWrapper.Select() to match, which keeps both implementations consistent through the interface.
That said, in practice EditTextBox is always a TextEditorWrapper in the Find/Replace flow, so this wasn't causing observable misbehavior. The caret was already positioned correctly, and I could not replicate this bug. The change is defensive, but the correct one to make. I re-tested some complex regex scenarios and it still looks good.

> Yes, Find/replace is more complicated that it sounds ;)

Yes, it is... it can easily turn into a never-ending patching marathon, especially that we’re baby-sitting Avalone internals here like carets and whatnot. At least I can break it with advanced regexps only these days so it has come a long way.

File: src/ui/Features/Main/MainViewModel.cs [hunk 1]
```diff
@@ -9568,9 +9568,7 @@ public async Task HandleFindResult(FindViewModel result)
                     EditTextBox.Text = subtitle.Text;
                 }
 
-                EditTextBox.CaretIndex = _findService.CurrentTextIndex;
-                EditTextBox.SelectionStart = _findService.CurrentTextIndex;
-                EditTextBox.SelectionEnd = _findService.CurrentTextIndex + _findService.CurrentTextFound.Length;
+                EditTextBox.Select(_findService.CurrentTextIndex, _findService.CurrentTextFound.Length);
```
[the remark this answers is not in the document]
Comment (mjuhasz) [reply src/ui/Features/Main/MainViewModel.cs:9571]: Fixed in df04c8d

File: src/ui/Features/Main/MainViewModel.cs [hunk 2]
```diff
@@ -9703,17 +9696,12 @@ private async Task FindPrevious()
...
-            if (EditTextBox.Text == string.Empty)
+            if (EditTextBox.Text != subtitle.Text)
             {
                 EditTextBox.Text = subtitle.Text;
             }
 
-            EditTextBox.CaretIndex = foundIndex;
-            EditTextBox.SelectionStart = foundIndex;
-            EditTextBox.SelectionEnd = foundIndex + foundText.Length;
+            EditTextBox.Select(foundIndex, foundText.Length);
```
[the remark this answers is not in the document]
Comment (mjuhasz) [reply src/ui/Features/Main/MainViewModel.cs:9704]: Fixed in df04c8d

File: src/ui/Features/Main/MainViewModel.cs [hunk 3]
```diff
@@ -9976,9 +9964,7 @@ public async Task HandleReplaceResult(ReplaceViewModel result)
                     EditTextBox.Text = subtitle.Text;
                 }
 
-                EditTextBox.CaretIndex = foundIndex;
-                EditTextBox.SelectionStart = foundIndex;
-                EditTextBox.SelectionEnd = foundIndex + foundText.Length;
+                EditTextBox.Select(foundIndex, foundText.Length);
```
[the remark this answers is not in the document]
Comment (mjuhasz) [reply src/ui/Features/Main/MainViewModel.cs:9967]: Same thing, fixed in df04c8d

File: src/ui/Features/Main/MainViewModel.cs [hunk 4]
```diff
@@ -9631,17 +9629,12 @@ private async Task FindNext()
...
-            if (EditTextBox.Text == string.Empty)
+            if (EditTextBox.Text != subtitle.Text)
             {
                 EditTextBox.Text = subtitle.Text;
             }
 
-            EditTextBox.CaretIndex = foundIndex;
-            EditTextBox.SelectionStart = foundIndex;
-            EditTextBox.SelectionEnd = foundIndex + foundText.Length;
+            EditTextBox.Select(foundIndex, foundText.Length);
```
[the remark this answers is not in the document]
Comment (mjuhasz) [reply src/ui/Features/Main/MainViewModel.cs:9637]: Also addressed in df04c8d
